src/pydvl/valuation/methods/knn_shapley.py

```diff
@@ -13,3 +13,143 @@
     the VLDB Endowment, Vol. 12, No. 11, pp. 1610–1623.
 
 """
+from __future__ import annotations
+
+import numpy as np
+from joblib import Parallel, delayed
+from numpy.typing import NDArray
+from sklearn.neighbors import NearestNeighbors
+from tqdm.auto import tqdm
+from typing_extensions import Self
+
+from pydvl.utils.status import Status
+from pydvl.valuation.base import Valuation
+from pydvl.valuation.dataset import Dataset
+from pydvl.valuation.result import ValuationResult
+from pydvl.valuation.utility import KNNClassifierUtility
+
+
+class KNNShapleyValuation(Valuation):
+    """Computes exact Shapley values for a KNN classifier.
+
+    This implements the method described in
+    (Jia, R. et al., 2019)<sup><a href="#jia_efficient_2019a">1</a></sup>.
+    It exploits the local structure of K-Nearest Neighbours to reduce the number
+    of calls to the utility function to a constant number per index, thus
+    reducing computation time to $O(n)$.
+
+    Args:
+        utility: KNNUtility with a KNN model to extract parameters from. The object
+            will not be modified nor used other than to call [get_params()](
+            <https://scikit-learn.org/stable/modules/generated/sklearn.base.BaseEstimator.html#sklearn.base.BaseEstimator.get_params>)
+        progress: Whether to display a progress bar.
+
+    """
+
+    def __init__(self, utility: KNNClassifierUtility, progress: bool = True):
+        self.utility = utility
+        self.progress = progress
+
+        config = self.utility.model.get_params()
+        self.n_neighbors = config["n_neighbors"]
+
+        del config["weights"]
+        self.helper_model = NearestNeighbors(**config)
+
+    def fit(self, data: Dataset) -> Self:
+        """Calculate exact shapley values for a KNN model on a dataset.
+
+        This fit method bypasses direct evaluations of the utility function and
+        calculates the Shapley values directly.
+
+        In contrast to other data valuation models, the runtime increases linearly
+        with the size of the test dataset.
+
+        Calculating the least core valuation is a computationally expensive task that
```

Comment:
Looks like a copy-paste leftover (or is there any connection to least core?)

Reply:
Thanks!